### PR TITLE
	modified:   Makefile

### PR DESCRIPTION
**Pull request by Sidharthshanu**
modified:   src/error_handler.h
	modified:   src/logger.c
	modified:   src/simulations/simulate_device_error.c
	new file:   src/simulations/sleep.c

### CODE REVIEW

**Comment by Sidharthshanu** [comment]
This branch has added some device errors there can be a conflict in calling files as that is unclear what the root is

File: src/simulations/simulate_device_error.c [hunk 1]
```diff
@@ -5,15 +5,94 @@
...
+            else{
+                printf("unkown problem in the error creation");
+            }
+            close(fd);
+            break;
+        case 4:
+            printf("Simulating device access error DEVICE_ACCESS_FAILURE\n");
+            fd = open("build/example.lock", O_RDWR);
+            
+            int fd = open("example.lock", O_RDWR);
```
Comment (Advaitgaur004) [inline src/simulations/simulate_device_error.c:76]: The file "example.lock" is not present in the build file, and it has not been initialized or created anywhere else.

File: src/error_handler.h
Comment (Advaitgaur004) [inline src/error_handler.h:1]: Have to include "ETXTBSY", "LOCK_EX" ,"LOCK_UN", "LOCK_NB" 

code should look like this,

```
// File: src/error_handler.h
#ifndef ERROR_HANDLER_H
#define ERROR_HANDLER_H

#include <errno.h>   // Added for standard error codes
#include <fcntl.h>   // Added for file locking macros (LOCK_EX, LOCK_NB, LOCK_UN)

// Define ETXTBSY if not available
#ifndef ETXTBSY
#define ETXTBSY 26
#endif

// Define file locking macros if not available
#ifndef LOCK_EX
#define LOCK_EX 2
#endif

#ifndef LOCK_NB
#define LOCK_NB 4
#endif

#ifndef LOCK_UN
#define LOCK_UN 8
#endif

// Enum for error types
typedef enum {
    MEMORY_ERROR,
    FILE_ACCESS_ERROR,
    DEVICE_ERROR,
    NULL_ERROR,       // Added NULL_ERROR for null pointer dereference
    UNKNOWN_ERROR,
    TXT_BUSY,         // Added TXT_BUSY for text file busy error
    DEVICE_ERROR_ACCESS_FAILURE,
    DEVICE_BUSY       // Added DEVICE_BUSY for device busy state
} ErrorType;

// Function to handle errors
void handle_error(ErrorType type, const char *message, int error_code);

#endif // ERROR_HANDLER_H
```

File: Makefile [hunk 2]
```diff
@@ -27,6 +27,11 @@ simulate_file_error: $(SIM_DIR)/simulate_file_error.c $(SRC_FILES)
 
 simulate_device_error: $(SIM_DIR)/simulate_device_error.c $(SRC_FILES)
 	$(CC) $(CFLAGS) $(SIM_DIR)/simulate_device_error.c $(SRC_FILES) -o $(BUILD_DIR)/simulate_device_error
+	$(CC) $(SIM_DIR)/sleep.c -o $(BUILD_DIR)/sleep
+	touch $(BUILD_DIR)/access.txt
+	chmod -wx $(BUILD_DIR)/access.txt
```
Comment (Advaitgaur004) [inline Makefile:32]: In some environments (for example, in mine), -wx is not available; rather, it shows -r-r-r-r or something similar. So, I just changed this to.
`chmod 444 $(BUILD_DIR)/access.txt`

File: src/simulations/simulate_device_error.c [hunk 3]
```diff
@@ -5,15 +5,94 @@
...
+            
+            // Clean up
+            flock(fd, LOCK_UN);
+            close(fd);
+            break;
+        default:
+            fprintf(stderr, "Invalid error code. Use 1, 2, 3 or 4.\n");
+            return -1;
     }
-    close(fd);
```
Comment (Advaitgaur004) [inline src/simulations/simulate_device_error.c:17]: Why didn't you close the file descriptor in the end?

**Pull request merged**